### PR TITLE
Add `SetContextClassLoader` utility class

### PR DESCRIPTION
**Pull request by basil**
Promoting https://github.com/jenkinsci/ldap-plugin/blob/3731881c37d1c66f8f43101b7f8b0cdd5f1eba56/src/main/java/jenkins/security/plugins/ldap/SetContextClassLoader.java into a core utility class.

### Example usages

Production code: https://github.com/jenkinsci/xcode-plugin/pull/114
Test code: https://github.com/jenkinsci/jaxb-plugin/pull/34

### Proposed changelog entries

Introduced `SetContextClassLoader` utility class to assist in plugin class loading.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6575"></a>

### CODE REVIEW

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/SetContextClassLoader.java:62]: Once we're on Java 11, we could perhaps remove the `clazz` argument and get the calling class via the [`StackWalker`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/StackWalker.html) API.

**Comment by basil** [comment]
I think I am going leave this as draft/on-hold until we require Java 11. I would love to have a clean API from day 1 (with zero arguments), but that only seems possible with Java 11 features like `StackWalker`. I would rather avoid introducing multiple one-argument constructors only to have to deprecate them later.

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (jglick) [reply core/src/main/java/jenkins/util/SetContextClassLoader.java:62]: FWIW I would much prefer the explicit `Class` argument as currently written here. It is hardly bothersome to pass, and sometimes you may actually want something other than the calling class: a class from a library, a plugin you depend upon, etc. Magically inferring an argument according to the caller is a poor practice in terms of legibility and predictability; for example, it means that supposedly safe refactorings like pull up method may quietly lead to different behavior.

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (jglick) [inline core/src/main/java/jenkins/util/SetContextClassLoader.java:61]: or just

```java
this(clazz.getClassLoader());
```

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (basil) [reply core/src/main/java/jenkins/util/SetContextClassLoader.java]: Well you can have your opinion and I can have mine. I actually think a no-argument version is preferable since you can cut and paste code from one class to another without having to worry about updating the calling class argument to reflect the new home of the code.

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (basil) [reply core/src/main/java/jenkins/util/SetContextClassLoader.java]: done

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (timja) [inline core/src/main/java/jenkins/util/SetContextClassLoader.java:65]: cool can hopefully simplify a few APIs with this, e.g. https://github.com/jenkinsci/configuration-as-code-plugin/blob/e345fd8df353db1a9fbd930968b87fce30548438/test-harness/src/main/java/io/jenkins/plugins/casc/misc/Util.java#L172

**Review by timja — APPROVED** [review]

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (jglick) [inline core/src/main/java/jenkins/util/SetContextClassLoader.java:65]: Still -0 on including the no-arg overload as in https://github.com/jenkinsci/jenkins/pull/6575#discussion_r877087840: `StackWalker` and similar idioms should be avoided in production code.

**Review by daniel-beck — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: core/src/main/java/jenkins/util/SetContextClassLoader.java
Comment (jglick) [reply core/src/main/java/jenkins/util/SetContextClassLoader.java:65]: (an `Object` argument delegating to the `Class` overload would be fine, for calls from non-`static` contexts)

**Comment by jglick** [comment]
BTW there are a few internal places that could use the new utility as a demonstration: `ClassicPluginStrategy`, `ContextResettingExecutorService`, `TimerTest`, `PluginManagerTest`

**Comment by basil** [comment]
Two examples of usages of the new utility were provided in the PR description as a demonstration. There are a lot more that could be done.

**Pull request merged**
